python/tests/test_highlevel.py

```diff
@@ -2141,6 +2141,20 @@ def test_num_children(self):
         for u in tree.nodes():
             assert tree.num_children(u) == len(tree.children(u))
 
+    def test_root_properties(self):
+        for ts in get_example_tree_sequences():
+            for tree in ts.trees():
+                if tree.has_single_root:
+                    assert tree.num_roots == 1
+                    assert tree.root != tskit.NULL
+                elif tree.has_multiple_roots:
+                    assert tree.num_roots > 1
+                    with pytest.raises(ValueError, match="More than one root exists"):
+                        _ = tree.root
+                else:
+                    assert tree.num_roots == 0
+                    assert tree.root == tskit.NULL
```

Comment:
Can we assert that all three cases are triggered?

Reply:
Oh yes, I was actually wondering how to do that. I've done it in the latest commit with a `set()`

Reply:
Oh, that was a good call. Turns out get_example_tree_sequences() doesn't return a completely empty ts. When it does, it reveals a lot of bugs. I'll try fixing some now, in a different PR.